dali/operators/generic/reduce/reduce.h

```diff
@@ -28,16 +28,51 @@
 #define REDUCE_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float)  // NOLINT
 
 namespace dali {
+namespace detail {
+
+class AxesHelper {
+ public:
+  explicit inline AxesHelper(const OpSpec &spec) {
+    has_axes_arg_ = spec.TryGetRepeatedArgument(axes_, "axes");
+    has_axis_names_arg_ = spec.TryGetArgument(axis_names_, "axis_names");
+    has_empty_axes_arg_ =
+      (has_axes_arg_ && (axes_.size() == 0)) || (has_axis_names_arg_ && axis_names_.empty());
+
+    DALI_ENFORCE(!has_axes_arg_ || !has_axis_names_arg_,
+      "Arguments `axes` and `axis_names` are mutually exclusive");
+  }
+
+  void PrepareAxes(const TensorLayout &layout, int sample_dim) {
+    if (has_axis_names_arg_) {
+      axes_ = GetDimIndices(layout, axis_names_).to_vector();
+      return;
+    }
+
+    if (!has_axes_arg_) {
+      axes_.resize(sample_dim);
+      std::iota(axes_.begin(), axes_.end(), 0);
+    }
+  }
+
+  bool has_axes_arg_;
+  bool has_axis_names_arg_;
+  bool has_empty_axes_arg_;
+  vector<int> axes_;
+  TensorLayout axis_names_;
+};
+
+}  // namespace detail
+
 template <
   template <typename T, typename R> class ReductionType,
   typename Backend,
   template <template <typename X, typename Y> class RType, typename BType> class ImplType>
-class Reduce : public Operator<Backend> {
+class Reduce : public Operator<Backend>, detail::AxesHelper {
```

Comment:
I got inspired by the reduce kernels. They are sort of composed through inheritance from various parts. Since it works and results in less code, I'll leave it as it is.

Reply:
I have the same suggestion. But not a must.